### DISCUSSION TITLE
Better PEC on embedded boundary

### DISCUSSION THREAD

**Discussion by kli-jfp** [BLAST-WarpX/warpx#5166] [Ideas]
Hi! I have noticed that the embedded boundary has issues with pec electromagnetic-solver boundary conditions. Since the outside electric field values are simply skipped, the electric field at the embedded boundary will be "imperfectly" reflected. When using a cylindrical embedded boundary this issue can cause constant m=4,8,12... modes in the em-fields. 

The solution (one atleast) to this is similar to the pec boundaries used on the 3D cartesian grid already in use.
1. You find the points outside of the embedded boundary within a certain distance (perhaps 2-3 times cell distance). 
2. Find the closest point on the embedded boundary
3. Find the point on the opposite side ("opposite point") of the embedded boundary
4. Interpolate the electric field in the "opposite point"
5. Mirror electric field from point 4 relative to the embedded boundary
6. Take the mirrored electric field from point 5 and set the electric field value in the outside points from step 1.

2D python toy model: red points outside polygon (embedded boundary) and green points are the "opposite points"
<img src="https://github.com/user-attachments/assets/e198a9f8-04d9-4c17-aa80-91ea7325efbb" width="300" />

3D python toy model: with lathe (2pi swept polygon)
<img src="https://github.com/user-attachments/assets/f05039b4-7011-45d2-aba6-536a25cf06e4" width="300" />

What is already implemented (as I can tell):
- AMReX creates an embedded boundary with EB2 class
- edge lengths and distance to eb etc. is already calculated in the code

What is missing (?):
- Find closest point to the embedded boundary given an outside point

Some questions that need to be answered:
1. Is the distance_to_eb the distance to the actual boundary (i.e. the analytical function that is parsed) or to the closest eb facet?
2. Will it use too much GPU memory to store all "opposite points", and create an array  that links them to their respective outside point?
3. Can the search for "opposite point" be made on-the-fly with GPU threads? If it can be computed quickly then there is no need to store any more information. This would have to be done after every EMsolver step for every point on the outside (within a distance) of the embedded boundary. 
4. The electric field values will also have to be interpolated to the "opposite points" which do not necessarily  lie on the cartesian grid. Then they must be mirrored etc. All this must be done every time the EM fields are solved.

Any thoughts? Would this be a feature that WarpX could benefit from? Also if anyone is a master at AMReX that can point in the right direction, please do!

**Comment by RemiLehe** [comment]
Thanks for suggesting this. This sounds like a useful feature.
Regarding the solution that you proposed: is this a well-known method, and if yes, could you point us to a more detailed description in the literature, so that we can understand the background?
Thanks,
